tree/ntuple/v7/src/RPageSinkBuf.cxx

```diff
@@ -0,0 +1,74 @@
+/// \file RPageSinkBuf.cxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \author Max Orok <maxwellorok@gmail.com>
+/// \date 2021-03-17
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RNTupleOptions.hxx>
+#include <ROOT/RNTupleModel.hxx>
+#include <ROOT/RPageSinkBuf.hxx>
+
+ROOT::Experimental::Detail::RPageSinkBuf::RPageSinkBuf(std::unique_ptr<RPageSink> inner)
+   : RPageSink(inner->GetNTupleName(), inner->GetWriteOptions()), fInner(std::move(inner)) {}
+
+void ROOT::Experimental::Detail::RPageSinkBuf::CreateImpl(const RNTupleModel &model)
+{
+   fBufferedColumns.resize(fLastColumnId);
+   fInnerModel = model.Clone();
+   fInner->Create(*fInnerModel);
+}
+
+ROOT::Experimental::RClusterDescriptor::RLocator
+ROOT::Experimental::Detail::RPageSinkBuf::CommitPageImpl(ColumnHandle_t columnHandle, const RPage &page)
+{
+   // TODO avoid frequent (de)allocations by holding on to allocated buffers in RColumnBuf
```

Comment:
We are currently allocating and deallocating buffered pages each time a cluster is committed, but we could hold on to allocated pages across cluster commits and only deallocate them when RPageSinkBuf is destroyed. This would need some care around tracking which pages are live, checking whether they could hold the requested amount of elements, etc.